### PR TITLE
Feature/global ens_metplus_climo_reorder_bugfix

### PR DESCRIPTION
**Pull request by StevenSimon-NOAA**
## Pull Request Testing ##

- [ ] Describe testing already performed for this Pull Request:</br>

After consulting with DTC, we found that none of the "active" verification metrics were affected by this global_ens metplus climo field ordering bug (alphabetical order in MetPlus 5.0.1 vs numerical order in MetPlus 6.0 Beta). Thus, the discrepancies of the stats files generated by MetPlus 5.0.1 vs MetPlus 6.0 Beta are solely a shuffle in the order of the stats each in each column. EVS v1 stats and plots processing are unaffected by this shuffle upon doing a comparative analysis of the stats files generated by the different versions of MetPlus. However, future iterations of EVS will incorporate BSS as a precipitation verification metric and thus would be impacted by this bug. Thus, the pertinent discrepancies in the stats files generated by the different versions of MetPlus do encompass discrepancies in BSS values. When this bugfix is implemented in the precip config files, those discrepancies in precip BSS are eliminated (same BSS values regardless of version of MetPlus).          

- [ ] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

1. Generate global_ens stats files using this bug fix. COMIN can be any global_ens prep achive.

2. Note that the global_ens prod stats files do not feature this bugfix and will be the "non-bugfix" control.

3. Compare the global_ens prod stats files vs the emc.vpppg MetPlus Beta stats files and note any discrepancies. The different ordering of the stats within each column is expected and does not affect the plots generated. However, there should be a discrepancy in BSS values for a given precip variable field in the precip stats files.

4. Compare the global_ens bugfix stats files vs the emc.vpppg MetPlus Beta stats files and note any discrepancies. The different ordering of the stats within each column is expected and does not affect the plots generated. However, there should not be a discrepancy in BSS values for a given precip variable field in the precip stats files.

5. To confirm that none of the active EVS v1 metrics are affected by this bug, you can do a similar comparison with global_ens prod plots and global_ens emc.vpppg MetPlus Beta plots. Those should be the same. Because BSS is not yet an active verification metric for precip, you will not be able to do a similar BSS plot comparison. So that will currently be limited to the stats file comparison discussed in 3) and 4).     

- [ ] Has the code been checked to ensure that no errors occur during the execution? **[Yes or No]**

Yes

- [ ] Do these updates/additions include sufficient testing updates? **[Yes or No]**

Maybe

- [ ] Please complete this pull request review by **[Fill in date]**.</br>

ASAP

## Pull Request Checklist ##

- [ ] Review the source issue metadata (required labels, projects, and milestone).
- [ ] Complete the PR description above.
- [ ] Ensure the PR title matches the feature branch name.
- [ ] Check the following:
- [ ]  Instructions provided on how to run
- [ ]  Developer's name is replaced by ${user} where necessary throughout the code
- [ ]  Check that the ecf file has all the proper definitions of variables
- [ ]  Check that the jobs file has all the proper settings of COMIN and COMOUT and other input variables
- [ ]  Check to see that the output directory structure is followed
- [ ]  Be sure that you are not using MET utilities outside the METplus wrapper structure

- [ ] After submitting the PR, select **Development** issue with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue.

### CODE REVIEW

File: parm/metplus_config/stats/global_ens/atmos_grid2grid/EnsembleStat_fcstGEFS_obsModelAnalysis_climoERA5.conf [hunk 1]
```diff
@@ -197,9 +198,9 @@ ENSEMBLE_STAT_OUTPUT_DIR = {OUTPUT_BASE}/stat/{model}
 #ENSEMBLE_STAT_CLIMO_STDEV_INPUT_DIR = {climpath}
 
 ENSEMBLE_STAT_CLIMO_MEAN_FILE_NAME = {climpath}/mean_{valid?fmt=%m%d}
-ENSEMBLE_STAT_CLIMO_MEAN_USE_FCST = True
+#ENSEMBLE_STAT_CLIMO_MEAN_USE_FCST = True
```
Comment (AliciaBentley-NOAA) [inline parm/metplus_config/stats/global_ens/atmos_grid2grid/EnsembleStat_fcstGEFS_obsModelAnalysis_climoERA5.conf]: @StevenSimon-NOAA Thank you for submitting this bugfix PR! Can you please remove these two new "commented out" lines? You can save "commented out" lines in your personal space for reference, but they can't be in the official submission to NCO. Thanks!

CC @PerryShafran-NOAA @ShelleyMelchior-NOAA

File: parm/metplus_config/stats/global_ens/atmos_grid2grid/EnsembleStat_fcstNAEFS_obsModelAnalysis_climoERA5.conf [hunk 2]
```diff
@@ -220,9 +221,9 @@ ENSEMBLE_STAT_OUTPUT_DIR = {OUTPUT_BASE}/stat/{model}
 #ENSEMBLE_STAT_CLIMO_STDEV_INPUT_DIR = {climpath}
 
 ENSEMBLE_STAT_CLIMO_MEAN_FILE_NAME = {climpath}/mean_{valid?fmt=%m%d}
-ENSEMBLE_STAT_CLIMO_MEAN_USE_FCST = True
+#ENSEMBLE_STAT_CLIMO_MEAN_USE_FCST = True
```
Comment (AliciaBentley-NOAA) [inline parm/metplus_config/stats/global_ens/atmos_grid2grid/EnsembleStat_fcstNAEFS_obsModelAnalysis_climoERA5.conf]: @StevenSimon-NOAA Similar to my previous comment, can these commented out lines please be removed from the PR? Thanks!

**Comment by AliciaBentley-NOAA** [comment]
@StevenSimon-NOAA Thank you for submitting this PR! Before it is merged, we need to clean up the new, commented out lines. You can save those commented out line for your own reference in a separate space, but we need to remove them from the official PR to keep things clean for NCO. Thanks!

CC @PerryShafran-NOAA @ShelleyMelchior-NOAA

**Comment by AliciaBentley-NOAA** [comment]
@StevenSimon-NOAA Thank you for removing the commented out lines! Next, we need you to revert back the changes you made to update PAST to LAST in this PR. That is outside the scope of the bugfix. We are not allowed to make that change is this PR. That is an EVS v2.0 update, since it will change the names of the files that NCO archive on HPSS and the files they push to emcrzdm for us. Can you please revert those changes back to how they currently are in ops EVS so that the only things changed in this PR relate to the climo order bug? Thank you!

CC @PerryShafran-NOAA @ShelleyMelchior-NOAA

File: parm/metplus_config/stats/mesoscale/snowfall/PcpCombine_fcstMESOSCALE_snow_vars.conf [hunk 3]
```diff
@@ -35,7 +35,7 @@ FCST_PCP_COMBINE_COMMAND = -subtract {ENV[job_file1]} 'name="{ENV[snow_var]}"; l
 FCST_PCP_COMBINE_INPUT_NAMES = {ENV[snow_var]}
 FCST_PCP_COMBINE_INPUT_ACCUMS = {ENV[input_accum]}
 FCST_PCP_COMBINE_INPUT_LEVELS = {ENV[input_level]}
-FCST_PCP_COMBINE_INPUT_OPTIONS = GRIB2_pdt=0;
+FCST_PCP_COMBINE_INPUT_OPTIONS = GRIB2_pdt={ENV[input_grib2_pdt]};
```
Comment (AliciaBentley-NOAA) [inline parm/metplus_config/stats/mesoscale/snowfall/PcpCombine_fcstMESOSCALE_snow_vars.conf]: @StevenSimon-NOAA Similarly to the nfcens updates getting into this PR and needing to be backed out, can this "mesoscale" update be backed out of your PR as well? These updates are intended for EVS v2.0 and not the EVS v1.0 bugfix. Thanks!

CC @PerryShafran-NOAA @ShelleyMelchior-NOAA

**Comment by StevenSimon-NOAA** [comment]
I think these revert updates are now all accounted for. Thanks for bringing these to my attention.

**Comment by AliciaBentley-NOAA** [comment]
@StevenSimon-NOAA Thank you, Steve! This PR looks good to me now! I appreciate your help with the bookkeeping!

CC @PerryShafran-NOAA @ShelleyMelchior-NOAA

**Comment by ShelleyMelchior-NOAA** [comment]
HI @StevenSimon-NOAA , I have cloned your fork and checked out branch `feature/global_ens_metplus_climo_reorder_bugfix`. I will get started on this soon.

**Comment by ShelleyMelchior-NOAA** [comment]
Do all drivers in `dev/drivers/scripts/stats/global_ens` need to be run to test the climo re-ordering? Or can a specific subset of drivers be run?

**Comment by StevenSimon-NOAA** [comment]
> Do all drivers in `dev/drivers/scripts/stats/global_ens` need to be run to test the climo re-ordering? Or can a specific subset of drivers be run?

That would just be the global_ens stats drivers for atmos g2g, precip, and snowfall. Any driver that falls into the g2g stats category.

**Comment by StevenSimon-NOAA** [comment]
I will be away from my computer for the rest of today. I will revisit this PR on Monday morning. Thanks!

**Comment by ShelleyMelchior-NOAA** [comment]
I have kicked off the drivers for all grid2grid, precip, and snowfall stats processing.
logfiles: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_global_ens_*.o*`
COMOUT=`/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0/stats/global_ens`
NOTE: at the time of this writing all processes have completed except for naefs, gefs, ecme, and cmce grid2grid.

Please review against EVSv1 in ops.

**Comment by StevenSimon-NOAA** [comment]
> I have kicked off the drivers for all grid2grid, precip, and snowfall stats processing. logfiles: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_global_ens_*.o*` COMOUT=`/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0/stats/global_ens` NOTE: at the time of this writing all processes have completed except for naefs, gefs, ecme, and cmce grid2grid.
> 
> Please review against EVSv1 in ops.

Upon review of the log files, I found no visceral ERRORs or WARNINGs. The WARNING found in the naefs precip log file is associated with a preexisting bugzilla to be addressed in EVS v2. All test small and large stats files are fully accounted for. Further review of the stats files revealed that only precip stats were affected by the climo ordering bug, specifically metrics like BSS. The other g2g atmos fields and their associated metrics values were unaffected. For precip BSS, the ops precip BSS vs the test precip BSS differ as expected for a given forecast lead and region. On the other hand, the beta precip BSS vs the test precip BSS match as expected for a given forecast lead and region.  

The reason why the other fields were unaffected was because the associated config files used an option that forces the climo dictionary to align with the forecast or obs dictionary no matter what, i.e, ENSEMBLE_STAT_CLIMO_MEAN_USE_FCST = True and GRID_STAT_CLIMO_MEAN_USE_FCST = True. Since the precip forecast dictionary had to be ordered explicitly, such an option could not be applied. Thus, an explicit reordering of the climo dictionary in the precip config files was required. Thus, the bugfix addresses the issue with the precip stats but does not affect the other g2g fields which do not have this issue.

**Comment by ShelleyMelchior-NOAA** [comment]
Thanks, Steve! Are there specific plots drivers that I should be running with these new stats?

Will the plots be meaningful w/ a single day of stats? Or should I run a few more days to at least get 3 data points for the plots?

**Comment by StevenSimon-NOAA** [comment]
> Thanks, Steve! Are there specific plots drivers that I should be running with these new stats?
> 
> Will the plots be meaningful w/ a single day of stats? Or should I run a few more days to at least get 3 data points for the plots?

That is not clear because precip BSS is not a statistic that is available to plot right now in EVS. That has not been merged into EVS yet. The BSS statistic is generated but it has not been added to the plotting code yet.

**Comment by ShelleyMelchior-NOAA** [comment]
OK, so can you determine what you need to with the PR testing so far? Is there anything else I need to do/run to determine the validity of the bugfix changes in this PR?

**Comment by StevenSimon-NOAA** [comment]
I think we should be good with what we have. Plus, we do not have the webpages to do plots comparisons. Otherwise, I would say create a few g2g and precip stats plots to demonstrate that the active metrics are unaffected by the climo ordering bug.

**Review by ShelleyMelchior-NOAA — APPROVED** [review]
Changes were tested and produced expected results. This PR is approved for  merge.

**Pull request merged**